In [1]:
from pycaret.classification import *
from pycaret.classification import load_experiment

In [2]:
import pandas as pd
import numpy as np
import pickle

from tqdm import tqdm
tqdm.pandas()

import os
from pprint import pprint
import re
from collections import defaultdict
from util import *
import copy

In [3]:
from datetime import datetime

time_begin = datetime.now()
print("Begining Time:\n", time_begin)

Begining Time:
 2024-02-25 18:56:16.566301


In [4]:
DATA_ADDRESS = "./data"
# os.listdir(DATA_ADDRESS)
# TRAIN_DIR = os.path.join(DATA_ADDRESS,'preprocessed','train')
TEST_DIR = os.path.join(DATA_ADDRESS,'preprocessed','test')

# os.listdir(TRAIN_DIR)

In [5]:
data_test = load_data(
    dir_feature = TEST_DIR,
    file_prefix = 'mfcc_len5_fft2048_mels128_mfcc17_',
    dir_df_index = os.path.join(DATA_ADDRESS,'df_index_test.pkl'),
    n_interval=500,
    flatten=True)

100%|██████████| 50/50 [00:00<00:00, 381.91it/s]


In [6]:
data_test

0          1          2          3          4          5  \
0     -77.335411 -76.044968 -75.780907 -76.207321 -76.968941 -78.054825   
1     -73.302490 -74.613892 -75.964531 -74.992409 -73.675034 -73.362015   
2     -70.028969 -70.884056 -71.527603 -70.559921 -69.526054 -68.636116   
3     -69.520584 -70.708282 -73.370262 -74.992615 -76.616226 -77.995964   
4     -71.714066 -68.675499 -67.467400 -66.962158 -67.511063 -67.628944   
...          ...        ...        ...        ...        ...        ...   
24795 -74.823402 -75.183197 -72.505684 -62.229679 -56.033108 -53.905876   
24796 -81.253174 -79.992989 -79.081581 -78.866364 -79.583725 -79.730759   
24797 -60.312725 -66.106125 -78.295441 -81.621445 -83.470802 -83.794319   
24798 -84.050369 -84.043938 -84.046204 -84.046669 -84.046188 -84.045700   
24799 -71.145798 -72.402138 -72.598503 -68.281807 -61.535854 -57.819702   

               6          7          8          9  ...      3663      3664  \
0     -78.889557 -79.136742 -78.707405 -77.364380  ...  1.666882  1.499124   
1     -74.261330 -75.696556 -76.656960 -77.399261  ...  0.682681  0.048326   
2     -67.744370 -66.134888 -65.494141 -67.076927  ... -0.003192  0.087003   
3     -78.455650 -79.094627 -80.521629 -80.936478  ...  0.023778  0.023779   
4     -67.616135 -67.188828 -67.699387 -67.895012  ...  1.495647  0.805673   
...          ...        ...        ...        ...  ...       ...       ...   
24795 -54.594402 -55.899620 -57.826359 -59.926170  ...  0.023778  0.023779   
24796 -80.617554 -81.187675 -81.978432 -81.488297  ...  0.290756  0.395896   
24797 -83.916412 -83.942101 -83.948776 -83.949234  ...  0.042555  0.023779   
24798 -83.349434 -79.142738 -68.420471 -62.473373  ...  0.042842 -0.031567   
24799 -56.814533 -56.517624 -57.455231 -59.650242  ...  0.023778  0.023779   

           3665      3666      3667      3668      3669      3670      3671  \
0      1.059202  0.374428 -0.468886 -1.169286 -0.854867 -0.729526 -0.158081   
1     -0.012902  0.545132  1.274721  1.936445  2.379305  1.697993  1.388240   
2      0.376900 -0.297952 -0.694625 -0.153152 -0.107342 -0.534249 -0.441920   
3      0.023779  0.023779  0.023779  0.023779  0.023779  0.023779  0.023779   
4      0.509418  0.759485  0.382243  0.110432 -0.015470 -0.084720  0.296677   
...         ...       ...       ...       ...       ...       ...       ...   
24795  0.023779  0.023779  0.023779  0.023779  0.023779  0.023779  0.023779   
24796  0.083245 -1.099505 -1.459450 -1.171437 -1.480871 -1.380765 -0.541722   
24797  0.036162  0.054693  0.106021  0.065853  0.037056  0.052584  0.061042   
24798 -0.049336 -0.035447  0.158722  0.485223  0.065013 -0.336305 -0.680135   
24799  0.023779  0.023779  0.023779  0.023779  0.023779  0.023779  0.023779   

       target  
0       192.0  
1       192.0  
2       192.0  
3       192.0  
4       192.0  
...       ...  
24795  2137.0  
24796  2137.0  
24797  2137.0  
24798  2137.0  
24799  2137.0  

[24800 rows x 3673 columns]

In [7]:
data_test.shape

(24800, 3673)

In [8]:
data_test.columns

Index([       0,        1,        2,        3,        4,        5,        6,
              7,        8,        9,
       ...
           3663,     3664,     3665,     3666,     3667,     3668,     3669,
           3670,     3671, 'target'],
      dtype='object', length=3673)

In [9]:
# logistic regression
clf_pre = load_model("model/test_source/logesticRegression")


Transformation Pipeline and Model Successfully Loaded


In [10]:
clf_pre

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(include=['0', '1', '2', '3', '4', '5', '6',
                                             '7', '8', '9', '10', '11', '12',
                                             '13', '14', '15', '16', '17', '18',
                                             '19', '20', '21', '22', '23', '24',
                                             '25', '26', '27', '28', '29', ...],
                                    transformer=SimpleImputer())),
                ('categorical_imputer',
                 TransformerWrapper(include=[],
                                    transformer=SimpleImputer(strategy='most_frequent'))),
                ('trained_model',
                 LogisticRegression(max_iter=1000, random_state=42))])

In [11]:
# def equal_error_rate(y_true, y_pred):
#     fpr, tpr, thresholds = roc_curve(y_true, y_pred)
#     eer = 1 - tpr[np.nanargmin(np.abs(fpr - (1 - tpr)))]
#     return eer

# def false_reject_rate(y_true, y_pred):
#     tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
#     return fn / (fn + tp)

def equal_error_rate(y_true, y_pred):
    fpr, tpr, thresholds = roc_curve(y_true, y_pred)
    eer = brentq(lambda x : 1. - x - interp1d(fpr, tpr)(x), 0., 1.)
    return eer


def false_reject_rate(y_true, y_pred):
    fpr, tpr, thresholds = roc_curve(y_true, y_pred)
    eer = brentq(lambda x : 1. - x - interp1d(fpr, tpr)(x), 0., 1.)
    thresh = interp1d(fpr, thresholds)(eer)
    frr = 1 - tpr[np.nanargmin(np.absolute(thresholds - thresh))]
    return frr

def evaluate_model(y_true,y_pred):
    dict_scores = {
        'eer' : equal_error_rate(y_true, y_pred)  
        ,'frr' : false_reject_rate(y_true, y_pred)
        ,'acc' : accuracy_score(y_true, y_pred)
        ,'recall' : recall_score(y_true, y_pred)
        ,'precision' : precision_score(y_true, y_pred)
        ,'f1' : f1_score(y_true, y_pred)
    }
    return dict_scores


In [12]:
# V1.1
from scipy.optimize import brentq
from scipy.interpolate import interp1d
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

dict_evaluation = defaultdict(dict)
# Step 1
unique_speakers = data_test['target'].unique()

for speaker_id in tqdm(unique_speakers):
    data_binary = data_test.copy()
    data_binary['target_binary'] = np.where(data_binary['target'] == speaker_id, 1, 0)
    
    # Step 2
    X = data_binary.drop(['target', 'target_binary'], axis=1)
    y = data_binary['target_binary']
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # Step 3
    clf = copy.deepcopy(clf_pre)
    clf.fit(X_train, y_train)
    
    # Step 4 - Evaluation
    y_pred_train = clf.predict(X_train)
    y_pred_test = clf.predict(X_test)

    dict_evaluation['train'][speaker_id] = evaluate_model(y_train,y_pred_train)
    dict_evaluation['test'][speaker_id] = evaluate_model(y_test,y_pred_test)
    



100%|██████████| 100/100 [1:04:28<00:00, 38.69s/it]


In [13]:
dict_evaluation

defaultdict(dict,
            {'train': {192.0: {'eer': 0.0,
               'frr': 0.0,
               'acc': 1.0,
               'recall': 1.0,
               'precision': 1.0,
               'f1': 1.0},
              5333.0: {'eer': 0.0,
               'frr': 0.0,
               'acc': 1.0,
               'recall': 1.0,
               'precision': 1.0,
               'f1': 1.0},
              606.0: {'eer': 0.0,
               'frr': 0.0,
               'acc': 1.0,
               'recall': 1.0,
               'precision': 1.0,
               'f1': 1.0},
              1079.0: {'eer': 0.0,
               'frr': 0.0,
               'acc': 1.0,
               'recall': 1.0,
               'precision': 1.0,
               'f1': 1.0},
              5029.0: {'eer': 0.0,
               'frr': 0.0,
               'acc': 1.0,
               'recall': 1.0,
               'precision': 1.0,
               'f1': 1.0},
              925.0: {'eer': 0.0,
               'frr': 0.0,
               'acc

In [14]:
# Calculate average evaluation scores for train and test
avg_scores = {
    'train': {metric: sum(dict_evaluation['train'][key][metric] for key in dict_evaluation['train']) / len(dict_evaluation['train']) for metric in dict_evaluation['train'][2137.0]},
    'test': {metric: sum(dict_evaluation['test'][key][metric] for key in dict_evaluation['test']) / len(dict_evaluation['test']) for metric in dict_evaluation['test'][2137.0]}
}

# Create DataFrame from the average scores
df = pd.DataFrame(avg_scores).T
df

eer       frr       acc    recall  precision        f1
train  0.000000  0.000000  1.000000  1.000000   1.000000  1.000000
test   0.053402  0.058007  0.999317  0.941993   0.989472  0.964707

In [15]:
time_end = datetime.now()
print("End Time:\n", time_end)
print("Run Time:\n", time_end-time_begin)

End Time:
 2024-02-25 20:00:47.393472
Run Time:
 1:04:30.827171


first out put

source no copy

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>eer</th>
      <th>frr</th>
      <th>acc</th>
      <th>recall</th>
      <th>precision</th>
      <th>f1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>train</th>
      <td>0.005000</td>
      <td>0.01</td>
      <td>0.999952</td>
      <td>0.99</td>
      <td>0.99</td>
      <td>0.99</td>
    </tr>
    <tr>
      <th>test</th>
      <td>0.400066</td>
      <td>NaN</td>
      <td>0.989742</td>
      <td>0.03</td>
      <td>0.03</td>
      <td>0.03</td>
    </tr>
  </tbody>
</table>
</div>

Second out put

all data

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>eer</th>
      <th>frr</th>
      <th>acc</th>
      <th>recall</th>
      <th>precision</th>
      <th>f1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>train</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>test</th>
      <td>0.053402</td>
      <td>0.058007</td>
      <td>0.999317</td>
      <td>0.941993</td>
      <td>0.989472</td>
      <td>0.964707</td>
    </tr>
  </tbody>
</table>
</div>